client/command/socks.go

```diff
@@ -0,0 +1,190 @@
+		return
+	}
+
+	var socksConnList []*net.TCPConn
+
+	// If session is closed then close the socks server
+	go func(implantSessionID uint32) {
+		// Continuously check if the connection to the implant that opened the socks server is still alive
+		for GetSession(fmt.Sprintf("%d", implantSessionID), rpc) != nil {
+		}
+
+		fmt.Println(Warn + " Closing all socks connections")
+		for _, conn := range socksConnList {
+			_ = conn.Close()
+		}
+
+		fmt.Println(Warn + " Killing socks server")
+		listener.Close()
+	}(session.ID)
+
+	for {
+		conn, err := listener.AcceptTCP()
+		if err != nil {
+			if strings.Contains(err.Error(), "use of closed network connection") {
+				break
+			} else {
+				fmt.Printf(Warn+"[0x09] %s\n", err)
+			}
+		}
+
+		socksConnList = append(socksConnList, conn)
+		go handleConnection(ctx, conn, rpc)
+	}
+
+	listener.Close()
+}
+
+func handleConnection(ctx *grumble.Context, conn *net.TCPConn, rpc rpcpb.SliverRPCClient) {
+	socksConn := new(SocksConn)
+	socksConn.ClientConn = conn
+
+	// Negotiate authentication
+	err := socksConn.HandleAuthRequest()
+	if err != nil {
+		fmt.Printf("[0x01] %s\n", err)
+		return
+	}
+
+	// Handle socks connection request
+	err = socksConn.HandleConnectRequest()
+	if err != nil {
+		fmt.Printf("[0x02] %s\n", err)
+		return
+	}
+
+	session := ActiveSession.Get()
+	if session == nil {
+		return
+	}
+
+	// Create an RPC tunnel, then start it before binding the shell to the newly created tunnel
+	rpcTunnel, err := rpc.CreateTunnel(context.Background(), &sliverpb.Tunnel{
+		SessionID: session.ID,
+	})
+	if err != nil {
+		fmt.Printf(Warn+"[0x03] %s\n", err)
+		return
+	}
+
+	// Start() takes an RPC tunnel and creates a local Reader/Writer tunnel object
+	tunnel := core.Tunnels.Start(rpcTunnel.TunnelID, rpcTunnel.SessionID)
+	if err != nil {
+		fmt.Printf(Warn+"[0x04] %s\n", err)
+		return
+	}
+
+	tcpTunnel, err := rpc.TCPTunnel(context.Background(), &sliverpb.TCPTunnelReq{
+		RemoteHost: socksConn.RemoteHost,
+		RemotePort: socksConn.RemotePort,
+		Request:    ActiveSession.Request(ctx),
+		TunnelID:   tunnel.ID,
+	})
+	if err != nil {
+		fmt.Printf(Warn+"[0x05] %s\n", err)
+		return
+	}
+
+	if tcpTunnel.StatusCode == 0x00 {
+		socksConn.ReturnSuccessConnectMessage()
+
+		cleanup := func() {
+			// Close the client socket
+			_ = socksConn.ClientConn.Close()
+
+			if core.Tunnels.Get(tunnel.ID) != nil {
+				// Send a message to close the tunnel
+				_, err := rpc.CloseTunnel(context.Background(), &sliverpb.Tunnel{
+					TunnelID: tunnel.ID,
+				})
+				if err != nil {
+					// {{if .Config.Debug}}
+					// fmt.Printf(Warn+"[0x06] %s\n", err)
+					// {{end}}
+					return
+				}
+			}
+		}
+
+		go func() {
+			for tunnel.IsOpen {
+				readArray := make([]byte, 1024)
+				bytesRead, err := tunnel.Read(readArray)
+				if err != nil {
+					cleanup()
+				} else if bytesRead != 0 {
+					_, err := socksConn.ClientConn.Write(readArray[:bytesRead])
+					// {{if .Config.Debug}}
```

Comment:
The templated `{{if .Config.Debug}}` is for implant side code, and is replaced at compile time. It has no effect on the client code, which mean these debug statements are always displayed on the client.
Please remove them before merging.

Reply:
Removed